hermes/bin/src/event/queue.rs

```diff
@@ -0,0 +1,178 @@
+//! Hermes event queue implementation.
+
+use std::{
+    sync::{
+        mpsc::{Receiver, Sender},
+        Arc,
+    },
+    thread::{self, JoinHandle},
+};
+
+use once_cell::sync::OnceCell;
+
+use super::{HermesEvent, HermesEventPayload, TargetApp, TargetModule};
+use crate::{
+    app::{HermesAppName, IndexedApps},
+    runtime_context::HermesRuntimeContext,
+    wasm::module::{Module, ModuleId},
+};
+
+/// Singleton instance of the Hermes event queue.
+static EVENT_QUEUE_INSTANCE: OnceCell<HermesEventQueue> = OnceCell::new();
+
+/// Hermes event queue error
+#[derive(thiserror::Error, Debug, Clone)]
+pub(crate) enum Error {
+    /// Target app not found.
+    #[error("Target app not found, app name: {0:?}.")]
+    AppNotFound(HermesAppName),
+
+    /// Target module not found.
+    #[error("Target module not found, module id: {0:?}.")]
+    ModuleNotFound(ModuleId),
+
+    /// Failed to add event into the event queue. Event queue is closed.
+    #[error("Failed to add event into the event queue. Event queue is closed.")]
+    CannotAddEvent,
+
+    /// Failed when event queue already been initialized.
+    #[error("Event queue already been initialized.")]
+    AlreadyInitialized,
+
+    /// Failed when event queue not been initialized.
+    #[error("Event queue not been initialized. Call `HermesEventQueue::init` first.")]
+    NotInitialized,
+}
+
+/// Hermes event queue execution loop thread handler
+pub(crate) type HermesEventLoopHandler = JoinHandle<anyhow::Result<()>>;
+
+/// Hermes event queue.
+/// It is a singleton struct.
+pub(crate) struct HermesEventQueue {
+    /// Hermes event queue sender
+    sender: Sender<HermesEvent>,
+}
+
+impl HermesEventQueue {
+    /// Creates a new instance of the `HermesEventQueue`.
+    /// Runs an event loop thread.
+    ///
+    /// # Errors:
+    /// - `Error::AlreadyInitialized`
+    pub(crate) fn init(indexed_apps: Arc<IndexedApps>) -> anyhow::Result<HermesEventLoopHandler> {
+        let (sender, receiver) = std::sync::mpsc::channel();
+
+        EVENT_QUEUE_INSTANCE
+            .set(Self { sender })
+            .map_err(|_| Error::AlreadyInitialized)?;
+
+        let event_loop = thread::spawn(move || event_execution_loop(&indexed_apps, receiver));
+
+        Ok(event_loop)
+    }
+
+    /// Get the singleton instance of the `HermesEventQueue`.
+    ///
+    /// # Errors:
+    /// - `Error::NotInitialized`
+    pub(crate) fn get_instance() -> anyhow::Result<&'static HermesEventQueue> {
```

Comment:
Don't expose OnceCell initialized data outside the module its created in.
Just make a single `add_event` function outside the `impl` and make it.
see suggested changes in PR to your branch.